### PR TITLE
the router implementation simplification

### PR DESCRIPTION
**Pull request by saledjenic**
This PR slightly changes the logic for calculating the best route based on the provided parameters.

To not diverge a lot from the current router logic (to avoid possibly hard-breaking changes) I've tried to improve the following things only:
- New file `router_v2.go` is added that way the logic of the current router is the same as it was.
- Some functions are here just cause we need them for the current router and we will remove them once we're sure that the new logic works fine. An example: we don't need `suggestedFees` function anymore, what it does is now covered with the new `getBaseFee` and `getPriorityFees` functions, but it's still there because of the current router.
- For calculating toke and bonder fees and amount out (an amount that will be received on the destination address) in the case of Hop bridge we don't use our logic (which was a try to replicate Hop's logic on our side which is a very complex one) anymore, but we call a dedicated API endpoint provided by Hop to get all those details.
- The old function `Can` from the `Bridge` interface is now named `AvailableFor` (also `Bridge` name for that interface is not the best one, we should rename it, but it dates from the time when it was used only for bridge txs, later other txs types were added).
- `Path` struct is updated with new property names:
```
type PathV2 struct {
	BridgeName     string
	From           *params.Network    // Source chain
	To             *params.Network    // Destination chain
	FromToken      *walletToken.Token // Token on the source chain
	AmountIn       *hexutil.Big       // Amount that will be sent from the source chain
	AmountInLocked bool               // Is the amount locked
	AmountOut      *hexutil.Big       // Amount that will be received on the destination chain

	SuggestedPriorityFees *PriorityFees // Suggested priority fees for the transaction

	TxBaseFee     *hexutil.Big // Base fee for the transaction
	TxPriorityFee *hexutil.Big // Priority fee for the transaction, by default we're using the Medium priority fee
	TxGasAmount   uint64       // Gas used for the transaction
	TxBonderFees  *hexutil.Big // Bonder fees for the transaction - used for Hop bridge
	TxTokenFees   *hexutil.Big // Token fees for the transaction - used for bridges (represent the difference between the amount in and the amount out)
	TxL1Fee       *hexutil.Big // L1 fee for the transaction - used for for transactions placed on L2 chains

	ApprovalRequired        bool            // Is approval required for the transaction
	ApprovalAmountRequired  *hexutil.Big    // Amount required for the approval transaction
	ApprovalContractAddress *common.Address // Address of the contract that needs to be approved
	ApprovalBaseFee         *hexutil.Big    // Base fee for the approval transaction
	ApprovalPriorityFee     *hexutil.Big    // Priority fee for the approval transaction
	ApprovalGasAmount       uint64          // Gas used for the approval transaction
	ApprovalL1Fee           *hexutil.Big    // L1 fee for the approval transaction - used for for transactions placed on L2 chains

	EstimatedTime TransactionEstimation

	requiredTokenBalance  *big.Int
	requiredNativeBalance *big.Int
}
```
- The router doesn't return any cost (via paths) in any currency, all returned values are in Wei (not GWei).
- The router builds candidates (PathV2) even if there is no balance on the “send from” address and after that it figures out what route (list of paths) is the best one based on the price and that's the route we will return to a client, but also we will check for the possibility of execution of such route and let the client know if the route can be executed based on balances for the selected address on each of the chains from the list.

Basically, those are the initial changes in this PR, I haven't tested them yet, but I wanted firstly to discuss with you is that ok and do we want to continue in that direction?

Since the old `Path` struct is changed and it corresponds to the new `PathV2` struct, that change is a breaking change cause it requires handling on the client side. But I see that now we have a more granular fee structure and it will be up to the client to shape it in the form it wants to display to a user.

That way we can cover complex routes like this:
image

image

### CODE REVIEW

**Comment by saledjenic** [comment]
Things that are not covered here be we should have them in mind are:
- update fees based on base fee and gas price change
- support swap via paraswap where we need to place an approval tx first and wait for its execution and then the swap transaction (that will be handled in `CreateMultiTransactionFromCommand` function)

File: services/wallet/bridge/hop.go [hunk 1]
```diff
@@ -109,27 +113,16 @@ func (h *HopBridge) Name() string {
 	return "Hop"
 }
 
-func (h *HopBridge) Can(from, to *params.Network, token *token.Token, toToken *token.Token, balance *big.Int) (bool, error) {
-	if balance.Cmp(big.NewInt(0)) == 0 {
-		return false, nil
-	}
-
+func (h *HopBridge) AvailableFor(from, to *params.Network, token *token.Token, toToken *token.Token) (bool, error) {
```
Comment (alaibe) [inline services/wallet/bridge/hop.go:116]: how do u check that hop supports the token?

File: services/wallet/bridge/hop.go [hunk 2]
```diff
@@ -349,142 +342,50 @@ func (h *HopBridge) swapAndSend(chainID uint64, hopArgs *HopTxArgs, signerFn bin
...
 
-// CalculateBonderFees logics come from: https://docs.hop.exchange/fee-calculation
-func (h *HopBridge) CalculateBonderFees(from, to *params.Network, token *token.Token, amountIn *big.Int, nativeTokenPrice, tokenPrice float64, gasPrice *big.Float) (*big.Int, error) {
-	amount := new(big.Float).SetInt(amountIn)
-	totalFee := big.NewFloat(0)
-	destinationTxFee, err := h.getDestinationTxFee(from, to, nativeTokenPrice, tokenPrice, gasPrice)
-	if err != nil {
-		return nil, err
-	}
+func (h *HopBridge) CalculateFees(from, to *params.Network, token *token.Token, amountIn *big.Int) (*big.Int, *big.Int, error) {
```
Comment (alaibe) [inline services/wallet/bridge/hop.go:345]: Those changes will also impact router v1, which is fine, did u ensure that it returns something consistent with [hop](https://hop.exchange/) ?

File: services/wallet/bridge/hop.go [hunk 3]
```diff
@@ -349,142 +342,50 @@ func (h *HopBridge) swapAndSend(chainID uint64, hopArgs *HopTxArgs, signerFn bin
...
+	)
 
-	bonderFeeRelative, err := h.getBonderFeeRelative(from, to, amount, token)
-	if err != nil {
-		return nil, err
+	fromChainName := HopMainnetChainName
+	if from.ChainID == walletCommon.OptimismMainnet {
+		fromChainName = HopOptimismChainName
+	} else if from.ChainID == walletCommon.ArbitrumMainnet {
+		fromChainName = HopArbitrumChainName
```
Comment (alaibe) [inline services/wallet/bridge/hop.go]: nit: make it a map and avoid if else if else if ;)

File: services/wallet/fees.go [hunk 4]
```diff
@@ -23,6 +24,9 @@ const (
 	GasFeeHigh
 )
 
+// //////////////////////////////////////////////////////////////////////////////
+// TODO: remove `SuggestedFees` struct once new router is in place
```
Comment (alaibe) [inline services/wallet/fees.go:29]: It is not only used by the router, so other places probably needs update

File: services/wallet/fees.go [hunk 5]
```diff
@@ -163,6 +176,57 @@ func (f *FeeManager) suggestedFees(ctx context.Context, chainID uint64) (*Sugges
 	}, nil
 }
 
+func (f *FeeManager) getBaseFee(ctx context.Context, client chain.ClientInterface, testnetMode bool) (*big.Int, error) {
```
Comment (alaibe) [inline services/wallet/fees.go]: Shall you update: suggestedFees in the meantime with the new function u created?

File: services/wallet/router.go [hunk 6]
```diff
@@ -694,22 +529,22 @@ func (r *Router) suggestedRoutes(
 						continue
 					}
 
-					if !sendType.isAvailableBetween(network, dest) {
```
Comment (alaibe) [inline services/wallet/router.go]: why removing this check?

File: services/wallet/router_send_type.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
+package wallet
```
Comment (alaibe) [inline services/wallet/router_send_type.go]: nit: create a router package?

File: services/wallet/router_v2.go [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+	}
+
+	return filteredRoutesLevel2
+}
+
+func findBestV2(routes [][]*PathV2, tokenPrice float64, nativeChainTokenPrice float64) []*PathV2 {
+	var best []*PathV2
+	bestCost := big.NewFloat(math.Inf(1))
+	for _, route := range routes {
+		currentCost := big.NewFloat(0)
```
Comment (alaibe) [inline services/wallet/router_v2.go]: Maybe add a cost function to the PathV2 Type?

File: services/wallet/router_v2.go [hunk 9]
```diff
@@ -0,0 +1,572 @@
+package wallet
```
Comment (alaibe) [inline services/wallet/router_v2.go]: is the logic of this file mostly a copy/paste from v1 adapted?

File: services/wallet/router_v2.go [hunk 10]
```diff
@@ -0,0 +1,572 @@
...
+				pathCost.Add(pathCost, new(big.Float).Mul(
+					new(big.Float).Quo(new(big.Float).SetInt(path.TxTokenFees.ToInt()), big.NewFloat(math.Pow(10, float64(path.FromToken.Decimals)))),
+					new(big.Float).SetFloat64(tokenPrice)))
+			}
+
+			if path.ApprovalRequired {
+				if path.ApprovalBaseFee != nil && path.ApprovalPriorityFee != nil {
+					feePerGas := new(big.Int).Add(path.ApprovalBaseFee.ToInt(), path.ApprovalPriorityFee.ToInt())
+					txFeeInWei := new(big.Int).Mul(feePerGas, big.NewInt(int64(path.ApprovalGasAmount)))
+					txFeeInEth := gweiToEth(weiToGwei(txFeeInWei))
```
Comment (alaibe) [inline services/wallet/router_v2.go]: why are u not returning fees directly in wei?

File: services/wallet/bridge/hop.go (hunk 1 above)
Comment (saledjenic) [reply services/wallet/bridge/hop.go:116]: That api we're calling `https://api.hop.exchange/v1/quote` will return an like this:
```
{
    "error": "Cannot destructure property 'name' of 'config_1.metadata.tokens[canonicalSymbol]' as it is undefined."
}
```
if the hope doesn't support selected token.

File: services/wallet/bridge/hop.go (hunk 2 above)
Comment (saledjenic) [reply services/wallet/bridge/hop.go:345]: They won't cause I removed these params `nativeTokenPrice, tokenPrice float64, gasPrice *big.Float` and they were used only for fees calculation for Hop, but now since we're calling hop API for that we don't need them. For other types that implement `Bridge` interface those params were already ignored.

File: services/wallet/bridge/hop.go (hunk 3 above)
Comment (saledjenic) [reply services/wallet/bridge/hop.go]: I agree, will do.

File: services/wallet/router_v2.go (hunk 9 above)
Comment (saledjenic) [reply services/wallet/router_v2.go]: What I did, I literally copied `router.go` file and then went through the router logic and reorganized some things taking care to not miss anything that we already have there. Guess this way is also easier to compare with the current router.

File: services/wallet/router_v2.go (hunk 8 above)
Comment (saledjenic) [reply services/wallet/router_v2.go]: I was thinking about it, but my intention was to avoid fetching prices when preparing candidate paths. 
We can easily do that anytime.

File: services/wallet/fees.go (hunk 4 above)
Comment (saledjenic) [reply services/wallet/fees.go:29]: Ahh I see, on the statusgo side it's used only by the router, but I see it's also exposed as an API point. 
Yes, that will require related updates on the client side for the parts that are using it (if those parts do not switch to using the router).

File: services/wallet/router_v2.go (hunk 10 above)
Comment (saledjenic) [reply services/wallet/router_v2.go]: All fees are in Wei and they are returned to the client in Wei via PathV2, but here we're converting them to `Eth` (Wei to GWei, then GWei to Eth) just to multiply the value with the Eth price to get the final price in Eth that we can compare in order to find the cheapest route.

File: services/wallet/router.go (hunk 6 above)
Comment (saledjenic) [reply services/wallet/router.go]: This check is implemented in `AvailableFor` function of each type individually.

File: services/wallet/fees.go (hunk 5 above)
Comment (saledjenic) [reply services/wallet/fees.go]: I counted that we will remove it sooner or later, but yes, will update it since this new one has a different chain config in the case of testnet.

**Comment by saledjenic** [comment]
@alaibe thanks for the review and your comments, they are appreciated since you're the most familiar with the router related code.

There is for sure space for further improvements here, but before we continue I would like to hear from others and be sure that we all have the same vision about it and aim in the same direction.

File: services/wallet/api.go [hunk 11]
```diff
@@ -445,6 +445,25 @@ func (api *API) GetSuggestedRoutes(
 		disabledToChaindIDs, preferedChainIDs, gasFeeMode, fromLockedAmount)
 }
 
+func (api *API) GetSuggestedRoutesV2(
+	ctx context.Context,
```
Comment (cammellos) [inline services/wallet/api.go]: if this is a public API, I would pass a request object (i.e a map), so that it's easier to identify which one is the parameter etc, minor thing though

File: services/wallet/bridge/hop.go (hunk 1 above)
Comment (alaibe) [reply services/wallet/bridge/hop.go:116]: alright it means it is handle when computing fees right?

File: services/wallet/bridge/hop.go (hunk 2 above)
Comment (alaibe) [reply services/wallet/bridge/hop.go:345]: I meant the body of the function, not the signature

File: services/wallet/bridge/hop.go (hunk 1 above)
Comment (saledjenic) [reply services/wallet/bridge/hop.go:116]: No, it's part of the router logic:
```
can, err := bridge.AvailableFor(network, dest, token, toToken)
if err != nil || !can {
	continue
}
```
means it Hop bridge doesn't support certain token we just skip the Hop bridge as potential candidate for the final route.

File: services/wallet/bridge/hop.go (hunk 2 above)
Comment (saledjenic) [reply services/wallet/bridge/hop.go:345]: At the end of that function it returns the token fee and bonder fee, the same as it was, we've just changed the way how we calculate them (now we're fetching them from the remote API). But all that will be undergone thorough checking.

File: services/wallet/bridge/cbridge.go [hunk 12]
```diff
@@ -5,9 +5,9 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	"io/ioutil"
 	"math/big"
-	"net/http"
+	netUrl "net/url"
```
Comment (Samyoul) [inline services/wallet/bridge/cbridge.go]: Why alias this import? I can't see a conflict with any other package import.

Comment (saledjenic) [reply services/wallet/bridge/cbridge.go]: You're absolutely right, while I was on it I had many other imports there and probably added because of some conflicts, but at the end when I cleaned up the code a bit, forgot to remove it.

File: services/wallet/bridge/cbridge.go [hunk 13]
```diff
@@ -60,43 +63,27 @@ func (s *CBridge) Name() string {
...
+	params.Add("src_chain_id", strconv.Itoa(int(from.ChainID)))
+	params.Add("dst_chain_id", strconv.Itoa(int(to.ChainID)))
+	params.Add("token_symbol", symbol)
+	params.Add("amt", amountIn.String())
+	params.Add("usr_addr", "0xaa47c83316edc05cf9ff7136296b026c5de7eccd")
+	params.Add("slippage_tolerance", "500")
 
-	resp, err := client.Do(req)
+	url := fmt.Sprintf("%s/v2/estimateAmt", base)
+	response, err := s.httpClient.DoGetRequest(context.Background(), url, params)
```
Comment (Samyoul) [inline services/wallet/bridge/cbridge.go:80]: Were you avoiding vars shadowing the `url` import with the `netUrl` alias? If yes, the convention (but not the rule) is to change the variable name, rather than the import alias.

Shadowing in Go is allowed, but many devs (myself included) avoid it like you have.

My opinion is that It is fine to make no changes here, just to point out that import aliasing is normally done to handle multiple package imports with the same name, and communicates that the dev should be careful when choosing which package they are selecting.

File: services/wallet/bridge/cbridge.go [hunk 14]
```diff
@@ -112,36 +99,18 @@ func (s *CBridge) getTransferConfig(isTest bool) (*cbridge.GetTransferConfigsRes
...
-		Timeout: time.Second * 5,
-	}
 	base := baseURL
 	if isTest {
 		base = testBaseURL
 	}
 	url := fmt.Sprintf("%s/v2/getTransferConfigs", base)
-
-	req, err := http.NewRequest(http.MethodGet, url, nil)
+	response, err := s.httpClient.DoGetRequest(context.Background(), url, nil)
```
Comment (Samyoul) [inline services/wallet/bridge/cbridge.go:107]: This is a nice change for the context.

File: services/wallet/bridge/hop.go
Comment (Samyoul) [inline services/wallet/bridge/hop.go:1]: Instead of deleting all the logic and depending entirely on hop's API, do you think that we could create a new function that handles the API calls? Just so we don't lose entirely the logic @alaibe implemented here.

We can even use the API calls to validate logic currently present.

File: services/wallet/bridge/swap_paraswap.go [hunk 15]
```diff
@@ -87,7 +87,7 @@ func (s *SwapParaswap) Can(from, to *params.Network, token *walletToken.Token, t
 	return true, nil
 }
 
-func (s *SwapParaswap) CalculateFees(from, to *params.Network, token *token.Token, amountIn *big.Int, nativeTokenPrice, tokenPrice float64, gasPrice *big.Float) (*big.Int, *big.Int, error) {
+func (s *SwapParaswap) CalculateFees(from, to *params.Network, token *token.Token, amountIn *big.Int) (*big.Int, *big.Int, error) {
```
Comment (Samyoul) [inline services/wallet/bridge/swap_paraswap.go:90]: Has pricing been removed because we use gwei as the common price?

File: services/wallet/fees.go [hunk 16]
```diff
@@ -163,6 +176,57 @@ func (f *FeeManager) suggestedFees(ctx context.Context, chainID uint64) (*Sugges
...
+		return priorityFee, err
+	}
+
+	suggestedPriorityFee, err := client.SuggestGasTipCap(ctx)
+	if err != nil {
+		return priorityFee, err
+	}
+
+	// Calculate Low priority fee
+	priorityFee.Low = fees[int64(0.1*float64(len(fees)))-1]
```
Comment (Samyoul) [inline services/wallet/fees.go:217]: How does counting the number of fees help determine the fee value?

`fees, err := f.getFeeHistorySorted(client.NetworkID())` gets historic fees sorted, but `len(fees)` just counts the number of fees, it doesn't handle the fee values.

File: services/wallet/router_send_type.go [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+	if s == ERC1155Transfer && b.Name() != ERC1155TransferString {
+		return false
+	}
+
+	if s != ERC1155Transfer && b.Name() == ERC1155TransferString {
+		return false
+	}
+
+	return true
+}
```
Comment (Samyoul) [inline services/wallet/router_send_type.go]: Could I make a couple of suggestions for this function:

1. Compatibility map approach:

Map each `SendType` to the expected bridge name. This way, the function looks up the required bridge name for a given `SendType` and compares it directly with the bridge’s name, this reduces the number of conditional statements.

```suggestion
func (s SendType) canUseBridge(b bridge.Bridge) bool {
    compatibilityMap := map[SendType]string{
        ERC721Transfer:  ERC721TransferString,
        ERC1155Transfer: ERC1155TransferString,
    }

    expectedBridge, ok := compatibilityMap[s]
    return !ok || b.Name() == expectedBridge
}
```

2. Switch approach

Directly map `SendType` to the condition without repeating the negative checks. 

```suggestion
func (s SendType) canUseBridge(b bridge.Bridge) bool {
    bridgeName := b.Name()
    switch s {
    case ERC721Transfer:
        return bridgeName == ERC721TransferString
    case ERC1155Transfer:
        return bridgeName == ERC1155TransferString
    default:
        return true // Assumes other types can use any bridge
    }
}
```

Both approaches have their advantages.

File: services/wallet/bridge/hop.go
Comment (saledjenic) [reply services/wallet/bridge/hop.go:1]: Yes, this is more a question for @alaibe, based on what he shared with me about what he was doing there, was a reverse engineering of Hop implementation and if they introduce any change to this logic will be hard for us to know about it. If we ever need it we can dig through git history.

File: services/wallet/bridge/swap_paraswap.go (hunk 15 above)
Comment (saledjenic) [reply services/wallet/bridge/swap_paraswap.go:90]: No, those params that are removed are not needed anymore, cause we needed them only to calculate bonder and token fee on our side, now we have them as a response of Hop's api.

File: services/wallet/fees.go (hunk 16 above)
Comment (saledjenic) [reply services/wallet/fees.go:217]: This logic was taken from `suggestedFees` function. I guess it's about 10th percentile, so we sort fees in an ascending way and take one that 10% of fees are under it, as the the low priority fee.
But if I'm mistaken @alaibe can explain better.

File: services/wallet/router_send_type.go [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	if network.ChainID == walletCommon.EthereumMainnet ||
+		network.ChainID == walletCommon.EthereumGoerli ||
+		network.ChainID == walletCommon.EthereumSepolia {
+		return true
+	}
+
+	return false
+}
```
Comment (Samyoul) [inline services/wallet/router_send_type.go]: This function has a number of compound conditionals, so perhaps we could declare the conditions first and then compare after. Something like:

```go
func (s SendType) isAvailableFor(network *params.Network) bool {
    // Set of network ChainIDs allowed for any type of transaction
    allAllowedNetworks := map[uint64]bool{
        walletCommon.EthereumMainnet: true,
        walletCommon.EthereumGoerli:  true,
        walletCommon.EthereumSepolia: true,
    }

    // Additional specific networks for the Swap SendType
    swapAllowedNetworks := map[uint64]bool{
        walletCommon.EthereumMainnet:  true,
        walletCommon.OptimismMainnet:  true,
        walletCommon.ArbitrumMainnet:  true,
    }

    // Check for Swap specific networks
    if s == Swap {
        return swapAllowedNetworks[network.ChainID]
    }

    // Check for any SendType available for all networks
    if s == Transfer || s == Bridge || s.IsCollectiblesTransfer() || allAllowedNetworks[network.ChainID] {
        return true
    }

    return false
}
```

This approach makes it more clear what the networks are used for what purpose.

File: services/wallet/api.go (hunk 11 above)
Comment (saledjenic) [reply services/wallet/api.go]: @cammellos just to confirm did you mean to have an API endpoint like this, where keys will be an upfront defined set of params or?
`func (api *API) GetSuggestedRoutesV2(ctx context.Context, params map[string]interface{})(*SuggestedRoutesV2, error)`

File: services/wallet/router_send_type.go [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+		return estimate
+	}
+
+	if s == ENSSetPubKey {
+		estimate, err := service.ens.API().SetPubKeyEstimate(context.Background(), network.ChainID, tx, fmt.Sprint(EstimateUsername, ".stateofus.eth"), EstimatePubKey)
+		if err != nil {
+			return 400000
+		}
+		return estimate
+	}
```
Comment (Samyoul) [inline services/wallet/router_send_type.go]: Perhaps this is a good case for a `switch` statement.

File: services/wallet/fees.go (hunk 16 above)
Comment (Samyoul) [reply services/wallet/fees.go:217]: Oh, I see. Yes, thank you.

File: services/wallet/api.go (hunk 11 above)
Comment (cammellos) [reply services/wallet/api.go]: not quite, I would use types:

```
type GetSuggestedRoutesV2Request struct {
  SendType SendType  `json:"sendType"`
  AddrFrom Commont.Address `json:"addrFrom"`
  ...
} 
```

Much the same as we do with `protocol/requests`
So it's immediately clear what this endpoints arguments are from the request and there's no risk of getting it wrong with positional parameters of the same type

File: services/wallet/bridge/hop.go (hunk 1 above)
Comment (alaibe) [reply services/wallet/bridge/hop.go:116]: if i am seeing correclty in bridge.AvailableFor for hop,  you are not looking at token anymore just network right?
if that is the case then the check for supported token is later? in fees computation?

File: services/wallet/router.go (hunk 6 above)
Comment (alaibe) [reply services/wallet/router.go]: After second thought:
```
func (s SendType) isAvailableBetween(from, to *params.Network) bool {
	if s.IsCollectiblesTransfer() {
		return from.ChainID == to.ChainID
	}

	if s == Bridge {
		return from.ChainID != to.ChainID
	}

	if s == Swap {
		return from.ChainID == to.ChainID
	}

	return true
}
```

This was use to prevent path to be created if send type prevent it. e.g u want to do a bridge only then you should only allow from and dest  that are different only. Are u still checking that, I cannot see it?

File: services/wallet/bridge/hop.go
Comment (alaibe) [reply services/wallet/bridge/hop.go:1]: My implementation was done before this endpoint existed, I wanted to switch to it so it is fine getting rid of it IMO

File: services/wallet/fees.go (hunk 16 above)
Comment (alaibe) [reply services/wallet/fees.go:217]: correct it is 10th percentile

File: services/wallet/bridge/hop.go (hunk 1 above)
Comment (saledjenic) [reply services/wallet/bridge/hop.go:116]: Correct, `CalculateFees` call will check that.

File: services/wallet/router.go (hunk 6 above)
Comment (saledjenic) [reply services/wallet/router.go]: Yes, all conditions you see here are actually covered and they are the first check in corresponding `AvailableFor` function implementations for each type that implements `Bridge` interface.

For example in case of Hop bridge that implementation looks like this:
```
func (h *HopBridge) AvailableFor(from, to *params.Network, token *token.Token, toToken *token.Token) (bool, error) {
	if from.ChainID == to.ChainID {
		return false, nil
	}
	.
	.
	.
	return true, nil
}
```

In case of Celar bridge looks like this:
```
func (s *CBridge) AvailableFor(from, to *params.Network, token *token.Token, toToken *token.Token) (bool, error) {
	if from.ChainID == to.ChainID {
		return false, nil
	}
	.
	.
	.
	return true, nil
}
```

File: services/wallet/api.go [hunk 20]
```diff
@@ -445,6 +445,25 @@ func (api *API) GetSuggestedRoutes(
...
+	sendType SendType,
+	addrFrom common.Address,
+	addrTo common.Address,
```
Comment (J-Son89) [inline services/wallet/api.go]: I see a lot of these parameters already exists so I don't know easy a change this is.
However, I find parameter/variable names that drop letters to be confusing. 
Sure we save a few characters but it's generally just better to use full words so the developer doesn't have to guess what the meaning is (not everyone is a native speaker) and on the otherside you almost have to guess what letters have been dropped as this is often an inconsistent and random choice. 
Overall it also leaves the codebase cleaner imo and that serves a lot in the long run. ⚔️

**Comment by saledjenic** [comment]
I addressed all your comments. If you're ok with that, we can merge these changes and I can continue inspecting each sending type individually and applying fixes if needed?

File: services/wallet/router/fees.go [hunk 21]
```diff
@@ -23,6 +25,9 @@ const (
 	GasFeeHigh
 )
 
+// //////////////////////////////////////////////////////////////////////////////
+// TODO: remove `SuggestedFees` struct once new router is in place
```
Comment (J-Son89) [inline services/wallet/router/fees.go:29]: can we include a GITHUB issue for this TODO - In Mobile that's how we have been doing it as it makes it easier to find out if the relevant issue was dropped or updated in any sense afterwards.

File: services/wallet/router/fees.go [hunk 22]
```diff
@@ -89,7 +100,13 @@ func gweiToWei(val *big.Float) *big.Int {
 	return res
 }
 
-func (f *FeeManager) suggestedFees(ctx context.Context, chainID uint64) (*SuggestedFees, error) {
+// //////////////////////////////////////////////////////////////////////////////
+// TODO: remove `suggestedFees` function once new router is in place
```
Comment (J-Son89) [inline services/wallet/router/fees.go:104]: same here

File: services/wallet/bridge/cbridge.go [hunk 23]
```diff
@@ -60,43 +63,27 @@ func (s *CBridge) Name() string {
...
-		symbol,
-		amountIn.String(),
-	)
 
-	req, err := http.NewRequest(http.MethodGet, url, nil)
-	if err != nil {
-		return nil, err
-	}
+	params := url.Values{}
+	params.Add("src_chain_id", strconv.Itoa(int(from.ChainID)))
```
Comment (J-Son89) [inline services/wallet/bridge/cbridge.go:72]: same for these, can't we just add full name params? :)
Why are some using this approach and some not? consistency adds a lot of value imo

Comment (saledjenic) [reply services/wallet/bridge/cbridge.go:72]: `src_chain_id ` this is not something that we defined, this is the param requested by clear API and it must be like that.

Comment (J-Son89) [reply services/wallet/bridge/cbridge.go:72]: Yep but the many cases below this too

**Review by alaibe — APPROVED** [review]
LGTM nice update

**Comment by briansztamfater** [comment]
> Since the old `Path` struct is changed and it corresponds to the new `PathV2` struct, that change is a breaking change cause it requires handling on the client side. But I see that now we have a more granular fee structure and it will be up to the client to shape it in the form it wants to display to a user.

Is there a simple way to see the diff between the old and new struct so it is easier to identify the changes to be made in the client side?

**Comment by saledjenic** [comment]
> > Since the old `Path` struct is changed and it corresponds to the new `PathV2` struct, that change is a breaking change cause it requires handling on the client side. But I see that now we have a more granular fee structure and it will be up to the client to shape it in the form it wants to display to a user.
> 
> Is there a simple way to see the diff between the old and new struct so it is easier to identify the changes to be made in the client side?

It will be explained how to migrate once we confirm that it works. 
The only changes for reponse will be in details about the fees.

File: services/wallet/router/router.go [hunk 24]
```diff
@@ -437,22 +277,41 @@ func newSuggestedRoutes(
...
-	erc1155Transfer := bridge.NewERC1155TransferBridge(s.rpcClient, s.transactor)
-	cbridge := bridge.NewCbridge(s.rpcClient, s.transactor, s.tokenManager)
-	hop := bridge.NewHopBridge(s.rpcClient, s.transactor, s.tokenManager)
-	paraswap := bridge.NewSwapParaswap(s.rpcClient, s.transactor, s.tokenManager)
+	transfer := bridge.NewTransferBridge(rpcClient, transactor)
+	erc721Transfer := bridge.NewERC721TransferBridge(rpcClient, transactor)
+	erc1155Transfer := bridge.NewERC1155TransferBridge(rpcClient, transactor)
+	cbridge := bridge.NewCbridge(rpcClient, transactor, tokenManager)
+	hop := bridge.NewHopBridge(rpcClient, transactor, tokenManager)
+	paraswap := bridge.NewSwapParaswap(rpcClient, transactor, tokenManager)
```
Comment (Samyoul) [inline services/wallet/router/router.go:288]: I'm sorry that I missed this previously. I notice that you've simplified the bridge inits by considerably increasing the `NewRouter()` init function's parameter complexity. I think that it is fine to pass in a single struct with multiple complex fields.

The Options pattern can help in this case or a config pattern. But also the original approach was fine. What problem does this change solve? Perhaps I'm not noticing something.

**Review by Samyoul — APPROVED** [review]
Thank you for the changes, and for all the work you've done to get this far. I've approved this PR with one minor point about function signature complexity (which is a point of opinion rather than a critical issue).

File: services/wallet/router/router.go (hunk 24 above)
Comment (saledjenic) [reply services/wallet/router/router.go:288]: @Samyoul That was an easy way to solve the cyclic dependency, that I ran into while moving the router to a new `router` package.

Without this the `router.go` (which is the `router` package) is importing the package `wallet` and `wallet/api.go` is importing `router` package while it's part of `wallet` package, that results in cyclic dependency.

Comment (Samyoul) [reply services/wallet/router/router.go:288]: Oh I see. Ok. I won't spend too much brain power on this then. Thank you.
